CoreImportExportPlugins/src/au/gov/asd/tac/constellation/plugins/importexport/svg/SVGGraph.java

```diff
@@ -0,0 +1,481 @@
+                    final SVGObject text = buildSVGObjectFromTemplate(SVGFileNameConstant.TOP_LABEL);
+                    final Float size = label.getSize() * 64;
+                    final Integer intSize = size.intValue();
+                    text.setAttribute("font-size", intSize.toString());
+                    text.setAttribute("y", String.format("%spx", offset.toString()));
+                    text.setAttribute(SVGAttributeConstant.FILL_COLOR.getKey(),label.getColor().getHtmlColor());
+                    text.setContent(SVGParser.sanitisePlanText(labelString));
+                    text.setParent(topLabelContainer.toSVGObject());
+                    offset = offset - intSize;
+                }
+            }
+        }
+        
+        /**
+         * Builds an image element within an SVG element.
+         * @param parent
+         * @param data 
+         */
+        private void buildSVGImageFromRasterImageData(final SVGObject parent, final byte[] data) {
+            final String encodedString = Base64.getEncoder().encodeToString(data);
+            final SVGObject image = buildSVGObjectFromTemplate(SVGFileNameConstant.IMAGE);
+            image.setAttribute("xlink:href", String.format("data:image/png;base64,%s", encodedString));
+            image.setParent(parent);
+        }
+        
+        /**
+         * Creates a SVGObject from a template SVG file
+         * The object will be returned with no parent.
+         * 
+         * @param templateResource the filename of the template file.
+         * @return 
+         */
+        private SVGObject buildSVGObjectFromTemplate(final SVGFileNameConstant templateResource) {
+            final InputStream inputStream = templateResource.getClass().getResourceAsStream(templateResource.resourceName);
+            SVGObject templateSVG = null;
+            try {
+                templateSVG = SVGParser.parse(inputStream);
+            } catch (final IOException ex) {
+                LOGGER.log(Level.INFO, Arrays.toString(ex.getStackTrace()));
+            }
+            return templateSVG;
+        }
+        
+        /**
+         * Creates a SVGGraph from a template SVG file.
+         * The object will be returned with no parent.
+         * Note an SVGGraph is a Wrapper around a SVGContainer.
+         * Note an SVGContainer is a wrapper around an SVGObject. 
+         * @param templateResource the filename of the template file.
+         * @return 
+         */
+        private SVGGraph buildSVGGraphFromTemplate(final SVGFileNameConstant templateResource) {
+            return new SVGGraph(new SVGContainer(buildSVGObjectFromTemplate(templateResource)));
+        }
+ 
+        /**
+         * Retrieves the min and max values for x and y for a given graph. 
+         * Represents the extremities of the graph area that contains render-able objects.
+         * @param graph
+         * @return 
+         */
+        private void defineBoundary(final GraphReadMethods graph) {
```

Comment:
Move this functionality to buildNodes() because you are looping through all nodes here a second time. Just calculate the bounds when you are initially looping through the nodes to get all the other attributes.

Reply:
The boundary needs to be defined before nodes are created as the SVG coordinates need to be adjusted from the Graph coordinates with respect to the boundary. The boundary needs to be known before this step. To simplify this method the `BoundingBox` and associated utilities have been used to calculate the extremities of the graph. If efficiency is an issue this bounding box could be improved to persist for the life of all graphs and be included in GraphReadMethods, it could then be updated dynamically as the graph is altered. Current implementations appear to generate a new `BoundingBox` as needed.